src/Functions/translate.cpp

```diff
@@ -0,0 +1,324 @@
+{
+    extern const int ILLEGAL_COLUMN;
+    extern const int ARGUMENT_OUT_OF_BOUND;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int BAD_ARGUMENTS;
+}
+
+struct TranslateImpl
+{
+    static void fillMapWithValues(
+        std::array<UInt8, 256> & map,
+        const std::string & map_from,
+        const std::string & map_to)
+    {
+        if (map_from.size() != map_to.size())
+            throw Exception("Second and trird arguments must be the same size", ErrorCodes::BAD_ARGUMENTS);
+
+        for (size_t i = 0; i < 256; ++i)
+            map[i] = i;
+
+        for (size_t i = 0; i < map_from.size(); ++i)
+        {
+            if (!isASCII(map_from[i]) || !isASCII(map_to[i]))
+                throw Exception("Second and trird arguments must be ASCII strings", ErrorCodes::BAD_ARGUMENTS);
+
+            map[map_from[i]] = map_to[i];
+        }
+    }
+
+    static void vector(
+        const ColumnString::Chars & data,
+        const ColumnString::Offsets & offsets,
+        const std::string & map_from,
+        const std::string & map_to,
+        ColumnString::Chars & res_data,
+        ColumnString::Offsets & res_offsets)
+    {
+        std::array<UInt8, 256> map;
+        fillMapWithValues(map, map_from, map_to);
+
+        res_data.resize(data.size());
+        res_offsets.assign(offsets);
+
+        UInt8 * dst = res_data.data();
+
+        for (UInt64 i = 0; i < offsets.size(); ++i)
+        {
+            const UInt8 * src = data.data() + offsets[i - 1];
+            const UInt8 * src_end = data.data() + offsets[i] - 1;
+
+            while (src < src_end)
+            {
+                *dst = map[*src];
+
+                ++src;
+                ++dst;
+            }
+
+            /// Technically '\0' can be mapped into other character,
+            ///  so we need to process '\0' delimiter separately
+            *dst++ = 0;
+        }
+    }
+
+    static void vectorFixed(
+        const ColumnString::Chars & data,
+        size_t /*n*/,
+        const std::string & map_from,
+        const std::string & map_to,
+        ColumnString::Chars & res_data)
+    {
+        std::array<UInt8, 256> map;
+        fillMapWithValues(map, map_from, map_to);
+
+        res_data.resize(data.size());
+
+        const UInt8 * src = data.data();
+        const UInt8 * src_end = data.data() + data.size();
+        UInt8 * dst = res_data.data();
+
+        while (src < src_end)
+        {
+            *dst = map[*src];
+
+            ++src;
+            ++dst;
+        }
+    }
+};
+
+struct TranslateUTF8Impl
+{
+    static void vector(
+        const ColumnString::Chars & data,
+        const ColumnString::Offsets & offsets,
+        const std::string & map_from,
+        const std::string & map_to,
+        ColumnString::Chars & res_data,
+        ColumnString::Offsets & res_offsets)
+    {
+        auto map_from_size = UTF8::countCodePoints(reinterpret_cast<const UInt8 *>(map_from.data()), map_from.size());
```

Comment:
Overall a nice and clean implementation. The only suggestion I have is that for the UTF8 case, we could have a fillMapWithValues() function like in the ASCII case.

Reply:
Oh, and translate[UTF8]() also needs documentation, e.g. here https://clickhouse.com/docs/en/sql-reference/functions/string-replace-functions